spec/client/ics-008-wasm-client/README.md

```diff
@@ -0,0 +1,447 @@
+
+### Motivation
+
+Wasm based light clients are decoupled from SDK source code, which enables one to upgrade existing light client or add support for other blockchain without modifying SDK source code.
+
+### Definitions
+
+Functions & terms are as defined in [ICS 2](../ics-002-client-semantics).
+
+`currentTimestamp` is as defined in [ICS 24](../ics-024-host-requirements).
+
+`Wasm Client Code` refers to Wasm bytecode stored in the client store, which provides a target blockchain specific implementation of [ICS 2](../ics-002-client-semantics).
+
+`Wasm Client` refers to a particular instance of `Wasm Client Code` defined as a tuple `(Wasm Client Code, ClientID)`.
+
+`Wasm VM` refers to a virtual machine capable of executing valid Wasm bytecode.
+
+
+### Desired Properties
+
+This specification must satisfy the client interface defined in ICS 2.
+
+## Technical Specification
+
+This specification depends on the correct instantiation of the `Wasm client` and is decoupled from any specific implementation of the target `blockchain` consensus algorithm.
+
+### Client state
+
+The Wasm client state tracks location of the Wasm bytecode via `codeId`. Binary data represented by `data` field is opaque and only interpreted by the Wasm Client Code. `type` represents client type.
+`type` and `codeId` both are immutable.
+
+```typescript
+interface ClientState {
+  codeId: []byte
+  data: []byte
+  latestHeight: Height
+  proofSpecs: []ProofSpec
+}
+```
+
+### Consensus state
+
+The Wasm consensus state tracks the timestamp (block time), `Wasm Client code` specific fields and commitment root for all previously verified consensus states.
+`type` and `codeId` both are immutable.
+
+```typescript
+interface ConsensusState {
+  codeId: []byte
+  data: []byte
+  timestamp: uint64
+}
+```
+
+### Height
+
+The height of a Wasm light client instance consists of two `uint64`s: the epoch number, and the height in the epoch.
+
+```typescript
+interface Height {
+  epochNumber: uint64
+  epochHeight: uint64
+}
+```
+
+Comparison between heights is implemented as follows:
+
+```typescript
+function compare(a: Height, b: Height): Ord {
+  if (a.epochNumber < b.epochNumber)
+    return LT
+  else if (a.epochNumber === b.epochNumber)
+    if (a.epochHeight < b.epochHeight)
+      return LT
+    else if (a.epochHeight === b.epochHeight)
+      return EQ
+  return GT
+}
+```
+
+This is designed to allow the height to reset to `0` while the epoch number increases by one in order to preserve timeouts through zero-height upgrades.
+
+### Headers
+
+Contents of Wasm client headers depend upon `Wasm Client Code`.
+
+```typescript
+interface Header {
+  data: []byte
+  height: Height
+}
+```
+
+### Misbehaviour
+
+The `Misbehaviour` type is used for detecting misbehaviour and freezing the client - to prevent further packet flow - if applicable.
+Wasm client `Misbehaviour` consists of two headers at the same height both of which the light client would have considered valid.
```

Comment:
Wasm clients may have other instances of misbehaviour other than two headers at the same height (time not monotonically increasing)? I think it is sufficient to say at minimum two headers are needed to prove misbehaviour?

Reply:
Fair point, this statement `at minimum two headers` makes sense to me. However, if we say "at a minimum" then would the `Misbehaviour` interface have to change? (to allow passing more than two headers?)

Reply:
Good point, lets just go with `Wasm client `Misbehaviour` consists of two conflicting headers both of which the light client would have considered valid.`? It's a little vague, but I don't think it is the responsibility of this spec to define what misbehaviour is exactly

Reply:
:+1:  this works for me, thanks